### PR TITLE
Handle the duplicate item issue by adding item in ddb conditionally

### PR DESCRIPTION
**Pull request by rakhimundhada15**
### Problem Description
O365 management api return the same message multiple time which result in duplicate incident . 

### Solution Description
Storing the logs for 24 hr and checking if the message is already exist return the exception else add to ddb table.
Once all  duplicate message is removed then send to ingest service.

### CODE REVIEW

**Review by imranalisyed506 — APPROVED** [review]
LGTM

File: paws_collector.js [hunk 1]
```diff
@@ -780,6 +808,79 @@ class PawsCollector extends AlAwsCollector {
...
+            const params = {
+                Item: {
+                    Id: { S: record[`${paramName}`] },
+                    CollectorId: { S: collector._collectorId },
+                    MsgHashId: { S: messageHashId },
+                    // setting DDB time to life. This is set to cover 24hr duplication window
+                    ExpireDate: { N: moment().add(1, 'days').unix().toString() }
+                },
+                TableName: collector._pawsDeDupLogsDdbTableName,
+                ConditionExpression: '(attribute_not_exists(Id) AND attribute_not_exists(MsgHashId)) OR  (Id = :id AND MsgHashId <> :msg_hash_id AND CollectorId = :collector_id)', // Check if item with same Id already exists and id and message is identical
```
Comment (scott-jones-AlertLogic) [inline paws_collector.js:845]: can we construct `Id` from `cid`, `collectorId`, and `messageHashId` and simplify the condition expression? the combination of cid + collector ID + messageHash should be enough to uniquely identify a message. we can still store other data for operational purposes though. 

maybe something like this?

```javascript
// ...
const cid = getCustomerId();  // I don't know where we usually take CID from 😅
const collectorId = collector._collectorId;
const messageHash = getHash(message);
const itemId = `${cid}_${collectorId}_${messageHash}`;
const params = {
  Item: {
    Id: { S: itemId },
    CollectorId: { S: collectorId },
    OrigMessageId: { S: record[`${paramName}`] },
    ExpireDate: { N: calculateExpiryTs() },
    TableName: collector._pawsDedDupLogsTableName,
    ConditionExpression: 'attribute_not_exists(Id)'
  }
};
// ...
```

File: paws_collector.js [hunk 2]
```diff
@@ -780,6 +808,79 @@ class PawsCollector extends AlAwsCollector {
...
+                        return resolve(null);
+                    }
+                });
+            });
+            promises.push(promise);
+        });
+
+        Promise.all(promises).then(result => {
+            if (duplicateCount > 0) {
+                collector.reportDuplicateLogCount(duplicateCount, (err)=>{
```
Comment (scott-jones-AlertLogic) [inline paws_collector.js:872]: cosmetic: please add spaces around `=>`

File: paws_collector.js [hunk 3]
```diff
@@ -780,6 +808,79 @@ class PawsCollector extends AlAwsCollector {
...
+                    if(err)
+                    AlLogger.warn(`PAWS000405 error from custom cloud watch metrics ${JSON.stringify(err)}`);
```
Comment (scott-jones-AlertLogic) [inline paws_collector.js:874]: this syntax scares me 😱 

could you please enclose the body of this expression in curly braces?
```javascript
if(err) {
  // ...
}
```

File: test/paws_mock.js [hunk 4]
```diff
@@ -117,6 +118,16 @@ const HEALTCHECK_SUBSCRIPTION_FILTERS = {
...
 };
+const MOCK_LOGS = [
+    {
+        "CreationTime": new Date().toISOString(), "Id": "c5d8e7ea-90b0-4549-9746-f67c8f6c00", "Operation": "UserLoginFailed", "Workload": "AzureActiveDirectory",
+        "ErrorNumber": "500121", "LogonError": "AuthenticationFailedSasError"
+    },
+    {
+        "CreationTime": new Date().toISOString(), "Id": "c5d8e7ea-90b0-4549-9746-f67c8f6c11", "Operation": "UserLoginFailed", "Workload": "AzureActiveDirectory",
+        "ErrorNumber": "500101", "LogonError": "AuthenticationFailedSasError"
+    }
```
Comment (scott-jones-AlertLogic) [inline test/paws_mock.js:129]: cosmetic: please format these in the usual JSON style with one property on each line 

```javascript
{
    CreationTime: new Date().toISOString(),
    Id: "c5d8e7ea-90b0-4549-9746-f67c8f6c00",
    // etc
}
```

File: paws_collector.js (hunk 3 above)
Comment (rakhimundhada15) [reply paws_collector.js:874]: sure

File: paws_collector.js [hunk 5]
```diff
@@ -780,6 +808,79 @@ class PawsCollector extends AlAwsCollector {
...
+        let duplicateCount = 0;
+        logs.forEach(record => {
+            const messageHashId = collector.getHash(record);
+            const params = {
+                Item: {
+                    Id: { S: record[`${paramName}`] },
+                    CollectorId: { S: collector._collectorId },
+                    MsgHashId: { S: messageHashId },
+                    // setting DDB time to life. This is set to cover 24hr duplication window
+                    ExpireDate: { N: moment().add(1, 'days').unix().toString() }
```
[the remark this answers is not in the document]
Comment (rakhimundhada15) [reply paws_collector.js:842]: Sure . I will add the condition base on env variable i.e `process.env.expire_ttl_hours`. I am not directly exposing the env variable in CFN template. Whenever required we can add in lambda env.

**Review by neil-bleasdale — APPROVED** [review]
This looks good to me. My comments are really just nit picking some minor formatting issues.

File: paws_collector.js [hunk 6]
```diff
@@ -780,6 +809,88 @@ class PawsCollector extends AlAwsCollector {
...
+     * Get hash of message 
+     * @param {*} message 
+     * @returns 
+     */
+    getHash(message) {
+        const hashValue = crypto.createHash('sha256').update(JSON.stringify(message, Object.keys(message).sort())).digest('hex');
+        return hashValue;
+    }
+
+/**
```
[the remark this answers is not in the document]
Comment (rakhimundhada15) [reply paws_collector.js:822]: i will check and update

**Review by scott-jones-AlertLogic — APPROVED** [review]

**Pull request merged**